src/main/java/gregtech/api/unification/material/Material.java

```diff
@@ -0,0 +1,848 @@
+package gregtech.api.unification.material;
+
+import com.google.common.base.CaseFormat;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import gregtech.api.unification.Element;
+import gregtech.api.unification.Elements;
+import gregtech.api.unification.material.info.MaterialFlag;
+import gregtech.api.unification.material.info.MaterialFlags;
+import gregtech.api.unification.material.info.MaterialIconSet;
+import gregtech.api.unification.material.properties.MaterialProperties;
+import gregtech.api.unification.material.properties.*;
+import gregtech.api.unification.stack.MaterialStack;
+import gregtech.api.util.SmallDigits;
+import net.minecraft.client.resources.I18n;
+import net.minecraft.enchantment.Enchantment;
+import net.minecraftforge.fluids.Fluid;
+import net.minecraftforge.fluids.FluidStack;
+import net.minecraftforge.fml.relauncher.Side;
+import net.minecraftforge.fml.relauncher.SideOnly;
+import stanhebben.zenscript.annotations.OperatorType;
+import stanhebben.zenscript.annotations.ZenOperator;
+
+import javax.annotation.Nonnull;
+import java.util.*;
+
+//@ZenClass("mods.gregtech.material.Material")
+//@ZenRegister
+public class Material implements Comparable<Material> {
+
+    /**
+     * Basic Info of this Material.
+     *
+     * @see MaterialInfo
+     */
+    @Nonnull
+    private final MaterialInfo materialInfo;
+
+    /**
+     * Properties of this Material.
+     *
+     * @see MaterialProperties
+     */
+    @Nonnull
+    private final MaterialProperties properties;
+
+    /**
+     * Generation flags of this material
+     *
+     * @see MaterialFlags
+     */
+    @Nonnull
+    private final MaterialFlags flags;
+
+    /**
+     * Chemical formula of this material
+     */
+    private String chemicalFormula;
+
+    // TODO Fix isotope tooltips being set toSmallDownNumbers
+    private String calculateChemicalFormula() {
+        if (chemicalFormula != null) return this.chemicalFormula;
+        if (materialInfo.element != null) {
+            return materialInfo.element.getSymbol();
+        }
+        if (!materialInfo.componentList.isEmpty()) {
+            StringBuilder components = new StringBuilder();
+            for (MaterialStack component : materialInfo.componentList)
+                components.append(component.toString());
+            return components.toString();
+        }
+        return "";
+    }
+
+    //@ZenGetter
+    public String getChemicalFormula() {
+        return chemicalFormula;
+    }
+
+    public Material setFormula(String formula) {
+        return setFormula(formula, false);
+    }
+
+    public Material setFormula(String formula, boolean withFormatting) {
+        this.chemicalFormula = withFormatting ? SmallDigits.toSmallDownNumbers(formula) : formula;
+        return this;
+    }
+
+    public ImmutableList<MaterialStack> getMaterialComponents() {
+        return materialInfo.componentList;
```

Comment:
I could see pack makers wanting to change composition of alloys, how about a setter for this?